src/libraries/System.Text.Json/tests/Serialization/CollectionTests/CollectionTests.KeyValuePair.cs

```diff
@@ -0,0 +1,438 @@
+        {
+            var kvp = new KeyValuePair<int, int>(1, 2);
+
+            JsonNamingPolicy namingPolicy = new TrailingAngleBracketPolicy();
+
+            // Baseline - properties serialized with default encoder if none specified.
+            var options = new JsonSerializerOptions
+            {
+                PropertyNamingPolicy = namingPolicy,
+            };
+
+            Assert.Equal(@"{""Key\u003C"":1,""Value\u003C"":2}", JsonSerializer.Serialize(kvp, options));
+
+            // Test - serializer honors custom encoder.
+            options = new JsonSerializerOptions
+            {
+                PropertyNamingPolicy = namingPolicy,
+                Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping
+            };
+
+            Assert.Equal(@"{""Key<"":1,""Value<"":2}", JsonSerializer.Serialize(kvp, options));
+        }
+
+        private class TrailingAngleBracketPolicy : JsonNamingPolicy
+        {
+            public override string ConvertName(string name) => name + "<";
+        }
+
+        [Theory]
+        [InlineData(typeof(KeyNameNullPolicy))]
+        [InlineData(typeof(ValueNameNullPolicy))]
+        public static void InvalidPropertyNameFail(Type policyType)
+        {
+            var options = new JsonSerializerOptions
+            {
+                PropertyNamingPolicy = (JsonNamingPolicy)Activator.CreateInstance(policyType)
+            };
+
+            InvalidOperationException ex = Assert.Throws<InvalidOperationException>(() => JsonSerializer.Deserialize<KeyValuePair<string, string>>("", options));
+            string exAsStr = ex.ToString();
+            Assert.Contains(policyType.ToString(), exAsStr);
+
+            Assert.Throws<InvalidOperationException>(() => JsonSerializer.Serialize(new KeyValuePair<string, string>("", ""), options));
+        }
+
+        private class KeyNameNullPolicy : JsonNamingPolicy
+        {
+            public override string ConvertName(string name) => name == "Key" ? null : name;
+        }
+
+        private class ValueNameNullPolicy : JsonNamingPolicy
+        {
+            public override string ConvertName(string name) => name == "Value" ? null : name;
+        }
+
+        [Theory]
+        [InlineData("")]
+        [InlineData("1")]
+        [InlineData("[")]
+        [InlineData("}")]
+        [InlineData("{")]
+        [InlineData("{}")]
+        [InlineData("{Key")]
+        [InlineData("{0")]
+        [InlineData(@"{""Random"":")]
+        [InlineData(@"{""Value"":1}")]
+        [InlineData(@"{""Value"":1,2")]
+        [InlineData(@"{""Value"":1,""Random"":")]
+        [InlineData(@"{""Key"":1,""Key"":1}")]
+        [InlineData(@"{""Key"":1,""Key"":2}")]
+        [InlineData(@"{""Value"":1,""Value"":1}")]
+        [InlineData(@"{""Value"":1,""Value"":2}")]
+        public static void InvalidJsonFail(string json)
+        {
+            Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<KeyValuePair<int, int>>(json));
```

Comment:
Is the JSON Path correct in these cases?

Reply:
We don't create `JsonPropertyInfo` instances for the key and value properties, so JSON path will not point directly to them, but to the property name for the KVP instance.

I think this granularity is good enough, but I'll estimate the effort to get the specific path then create an issue.

Reply:
Keep in mind that there is already an issue for general consistency of JSON Path https://github.com/dotnet/runtime/issues/31955, but sounds like you want to do something scoping only KVP, so maybe make sure to link them.